### DISCUSSION TITLE
Seeking some information on BVH optimization

### DISCUSSION THREAD

**Discussion by lmtss** [bepu/bepuphysics2#232] [Q&A] [answered]
Hi,

I'm new to physics engine and I'm learning optimization of bvh through your code. Could you provide some references about your implementation? Like some papers?

Tree_BinnedRefine.cs confuses me a bit, I wonder if there is a literature reference for this.

**Comment by RossNordby** [comment, marked as answer]
Binned BVH builders are described in this paper: https://www.sci.utah.edu/~wald/Publications/2007/ParallelBVHBuild/fastbuild.pdf

During a refinement step, the `Tree` applies binned builders to parts of a tree (not just subsets of leaves; blobs of internal nodes can be optimized too). It turns out that, over time, this converges to high quality with very low per-step cost.

There is some trickiness to deciding *where* to apply those refinements. The `BroadPhase` schedules them near the root and close to a subset of leaves stochastically chosen partially by increases in SAH cost that indicate the existing structure is becoming worse.

It's worth noting that the `Tree` implementation is *old*. By far the oldest part of bepuphysics2. A lot of it is based on assumptions and requirements that no longer exist, so don't be surprised if you find some pretty dang suboptimal stuff in there. It's on the todo list for a major overhaul. The new version will still be conceptually similar on a high level, though; I don't anticipate moving away from incremental binned optimization.

**Reply by lmtss** [reply]
Thank you! This is helpful !

**Reply by lmtss** [reply]
Do you have any new ideas now? I am looking for relevant information, such as this paper: https://meistdan.github.io/publications/prbvh/paper.pdf
Can I get some advice on the advanced BVH optimization？
I just finished reading that article([https://www.sci.utah.edu/~wald/Publications/2007/ParallelBVHBuild/fastbuild.pdf](https://www.sci.utah.edu/~wald/Publications/2007/ParallelBVHBuild/fastbuild.pdf)) and I think I learned a lot. Do you have your favorite latest paper on binned builders?

**Reply by RossNordby** [reply]
PRBVH is one of the more recent ones. Karras also had a paper on optimizing treelets: https://research.nvidia.com/sites/default/files/pubs/2013-07_Fast-Parallel-Construction/karras2013hpg_paper.pdf

There was another paper that I don't recall the name of that used smaller scale tree rotations to incrementally optimize as well.

A lot of the papers have GPU execution in mind and so generally push for implementations which scale widely extremely well. Scalability is still important for CPUs, but there's a big difference between 64 CPU cores and 10496 GPU "cores". CPU cores are much beefier and that can affect the choice of algorithm.

The binned refinement process in bepuphysics2 isn't based on any paper (and the future improvements aren't either), but the general idea is that larger scope subtree optimizations can give you a lot of bang for your buck on the CPU. Even with a dynamic scene, SAH cost remains very good and often comparable to sweep builders at a fraction of the cost. Technically, using local optimal refinements as in the above paper could push you to higher qualities than a sweep builder with sufficient computation, but it's very difficult to get *significantly* better. I'm not primarily concerned about improving SAH, since the maximum possible benefits there are quite small.

Instead, the revamp will be primarily concerned with performance. The implementation is fairly low quality, and each refinement region is sequential. This is a bit of a problem- there tends to be a bound on how fast the broad phase refinement step can be because even one refinement region will bottleneck it. It's still fast but giving large regions the ability to scale to more threads will make it much faster.

Another piece will be cache optimization. The current refinement process does not guarantee DFS ordering in memory. That's not actually a huge problem for the broad phase since the access pattern is a lot friendlier than a bunch of divergent rays, but it would still be nice to improve. Previously, there existed an explicit layout optimizer step, but the cost was comparable to the benefit, so I removed it. In the new version, I'd like to simply output the refined tree in DFS order directly.

**Reply by lmtss** [reply]
What do you think of the reinsert method? While it's hard to parallelize this way, PRBVH changes that. His method looks good for GPU parallelism, but what about CPU parallelism? I don't know how to evaluate this.
Is this the paper ([Tree Rotations for Improving Bounding Volume Hierarchies](http://eastfarthing.com/publications/tree.pdf)) you mentioned about tree rotation?
I just finished reading that article you mentioned([Fast Parallel Construction of High-Quality Bounding Volume Hierarchies](https://research.nvidia.com/sites/default/files/pubs/2013-07_Fast-Parallel-Construction/karras2013hpg_paper.pdf)). This is where the concept of treelet in your code comes from?
I found the BVH4 (4-way AABB tree) is a good idea. BVH4 can be combined with SIMD, like [titanfall](https://www.gdcvault.com/play/1025126/Extreme-SIMD-Optimized-Collision-Detection) and [Embree](https://www.embree.org/papers/2014-Siggraph-Embree.pdf), but it seems difficult to achieve. Have you considered changing BVH2 to BVH4?

**Reply by RossNordby** [reply]
>What do you think of the reinsert method? While it's hard to parallelize this way, PRBVH changes that.

Worth noting the build times on PRBVH. It's a huge improvement over the sequential reinsertion variant, but it's not targeting real time use. You might be able to get it to a workable speed by doing optimization over stochastic subsets, but I'm probably not going to investigate that.

>Is this the paper ([Tree Rotations for Improving Bounding Volume Hierarchies](http://eastfarthing.com/publications/tree.pdf)) you mentioned about tree rotation?

Looks like it.

>I just finished reading that article you mentioned([Fast Parallel Construction of High-Quality Bounding Volume Hierarchies](https://research.nvidia.com/sites/default/files/pubs/2013-07_Fast-Parallel-Construction/karras2013hpg_paper.pdf)). This is where the concept of treelet in your code comes from?

The concept of a treelet is older, and I suspect it's one of those things that's been independently thought up several times. If I remember correctly, I had been kicking the idea of binned incremental refinement around in my head for a while and seeing that paper's results solidified my intent to try it.

>I found the BVH4 (4-way AABB tree) is a good idea. BVH4 can be combined with SIMD, like [titanfall](https://www.gdcvault.com/play/1025126/Extreme-SIMD-Optimized-Collision-Detection) and [Embree](https://www.embree.org/papers/2014-Siggraph-Embree.pdf), but it seems difficult to achieve. Have you considered changing BVH2 to BVH4?

Considered it, yup. The best layout of nodes is sensitive to what you actually intend to do with those nodes, though. 

For example, if you know 1) you have 16-wide operations, 2) the tree is built offline, and 3) the use case is trying to accelerate a single ray test as much as possible, then using a 16-ary tree *could* be a good option (provided you handle the stack carefully).

Some of my internal projects have actually used up to 256-ary trees for certain types of GPU query. That made sense in context because the thread group was already 256 threads for other reasons, and in that design every thread depended on the result of one tree traversal. So, might as well throw every thread at the one traversal with one thread (SPMD lane) per child.

But the broad phase's main use of the tree is tree-tree tests. The number of necessary tests per node pair grows with the square of the arity, so a 256-ary tree would be horrible. It's *possible* that 4 or 8 children (or even 16 if you had AVX512) could still end up being a win thanks to the wider execution, but it's pretty questionable when you include the overhead associated with generating a higher arity tree. Spending more time on tree management to get a traversal cost reduction is often a net loss when traversal cost is already a tiny fraction of the frame.

Higher arities also come with sneaky complexity. For example, you have to be careful about how you combine nodes or else you could end up with large fractions of nodes being mostly empty. 

So I might still do a few more experiments, but I'm expecting to stick to binary trees for the `Tree` and broad phase.

**Reply by lmtss** [reply]
Hi，
I'm still learning from your code, but some problems arise.
According to the code (GetRefitAndMarkTuning):
$refinementLeafCountThreshold=min(N_{leaf}, 3\sqrt{N_{leaf}})$
and
```c#
// RefitAndMark
if (child.LeafCount <= leafCountThreshold)
{
    refinementCandidates.Add(child.Index, pool);
}
```
Does this mean that Refine can only optimize the lower layers of the tree, especially when there are many leaves ?
![未命名文件 (1)](https://user-images.githubusercontent.com/22532316/186065135-348e0fd3-15da-4cdf-881a-e2e34bc48ed6.png)

Is that so ？

**Reply by RossNordby** [reply]
Refinements are scheduled for a subset of subtrees near the leaves, yes, but also at the root. The root refinement extends far enough down that the entire tree is covered over time.

**Reply by lmtss** [reply]
My fault. I missed the code to add the root node to the refinement list. Thank you.

**Reply by lmtss** [reply]
I saw the comments in the BinnedResources struct. I would like to know what is a good way to optimize this preallocation. What is the zero-overhead method mentioned in the comments?

**Reply by RossNordby** [reply]
Hmm. This is why you don't write that sort of tantalizing comment without more details!

I don't exactly remember what I was referring to there. The 'overhead' associated with the current `BinnedResources` is already near zero in terms of runtime costs; it's just a bit gross looking.

If I had to guess, I was referring to major structural changes to the binned refiner which would make that kind of allocation less necessary, or at the very least different. Also, with module-level `SkipLocalsInit`, stackalloc doesn't zero memory. Given that the binned resources tend to scale with bin count, they would easily fit on the stack (with some care to avoid exploding things with a bad recursive implementation).

Or maybe I wrote that comment back when it was using a bunch of independent allocations instead suballocating.

I'm probably forgetting something :P

**Reply by lmtss** [reply]
Thank you!
>I was referring to major structural changes to the binned refiner which would make that kind of allocation less necessary

What kind of binned builder makes preallocation less necessary? I think the binned builder needs to pre-allocate some space to store the information of each bin (bin-bounds, bin-leafcount..) and the indexmap of the leaf nodes. I can only think of reducing some unnecessary caches, such as the precalculating centroids mentioned in the comments

**Reply by RossNordby** [reply]
I'd expect a fairly similar amount of memory to be used overall (within some constant factor), It's mostly just moving allocations around in a way that could be a bit more convenient for the implementation. `stackalloc` where possible and useful, for example. Possibly some changes for a better vectorized implementation.

**Reply by lmtss** [reply]
Thank you for your patience in answering my doubts.

**Reply by RossNordby** [reply]
Well, I got around to actually revamping the binned refinement process. Turns out there were a lot of opportunities for improvements. The only preallocations are now some data for bins: https://github.com/bepu/bepuphysics2/blob/32eb5b9fc6cb5dde2d9fb375dea85b2704fd64a9/BepuPhysics/Trees/Tree_BinnedBuilder.cs#L112-L121

It works on the input buffers in-place, so there's no need for additional resources.

It's also 4-6 times faster :P

(Not quite done; it's not actually used in the engine yet- going to implement a multithreaded codepath too.)

**Reply by lmtss** [reply]
Cool! This is a good reference!! Thank you!
